enzyme/Enzyme/MLIR/Interfaces/EnzymeLogicReverse.cpp

```diff
@@ -0,0 +1,190 @@
+        blocks.push_back(predecessorRevMode);
+        indices.push_back(APInt(32, i++));
+        arguments.push_back(ValueRange(operands));
+      }
+      else{
+        defaultBlock = predecessorRevMode;
+        defaultArguments = ValueRange(operands);
+      }
+    }
+    //Remove Dependency to CF dialect
+    if (std::next(oBB->getPredecessors().begin()) == oBB->getPredecessors().end()){
+      //If there is only one block we can directly create a branch for simplicity sake
+      revBuilder.create<cf::BranchOp>(gutils->getNewFromOriginal(&*(oBB->rbegin()))->getLoc(), defaultBlock, defaultArguments);
+    }
+    else{
+      Value cache = gutils->insertInitBackwardCache(gutils->getIndexCacheType());
+      Value flag = revBuilder.create<enzyme::PopCacheOp>(oBB->rbegin()->getLoc(), gutils->getIndexType(), cache);
+
+      revBuilder.create<cf::SwitchOp>(oBB->rbegin()->getLoc(), flag, defaultBlock, defaultArguments, ArrayRef<APInt>(indices), ArrayRef<Block *>(blocks), ArrayRef<ValueRange>(arguments));
+      
+      int j = 0;
+      for (Block * predecessor : oBB->getPredecessors()){
+        Block * newPredecessor = gutils->getNewFromOriginal(predecessor);
+        OpBuilder predecessorBuilder(newPredecessor, std::prev(newPredecessor->end()));
+
+        Value indicator = predecessorBuilder.create<arith::ConstantIntOp>(oBB->rbegin()->getLoc(), j++, 32);
+        predecessorBuilder.create<enzyme::PushCacheOp>(oBB->rbegin()->getLoc(), cache, indicator);
+      }
+    }
+  }
+}
+
+FunctionOpInterface mlir::enzyme::MEnzymeLogic::CreateReverseDiff(FunctionOpInterface fn, DIFFE_TYPE retType, std::vector<DIFFE_TYPE> constants, MTypeAnalysis &TA, bool returnUsed, DerivativeMode mode, bool freeMemory, size_t width, mlir::Type addedType, MFnTypeInfo type_args, std::vector<bool> volatile_args, void *augmented, SymbolTableCollection &symbolTable) {
+  
+  if (fn.getFunctionBody().empty()) {
+    llvm::errs() << fn << "\n";
+    llvm_unreachable("Differentiating empty function");
+  }
+
+  ReturnType returnValue = ReturnType::Tape;
+  MDiffeGradientUtilsReverse * gutils = MDiffeGradientUtilsReverse::CreateFromClone(*this, mode, width, fn, TA, type_args, retType, /*diffeReturnArg*/ true, constants, returnValue, addedType, symbolTable);
+
+
+  SmallVector<mlir::Block*> dominatorToposortBlocks = getDominatorToposort(gutils);
+
+  for (auto it = dominatorToposortBlocks.rbegin(); it != dominatorToposortBlocks.rend(); ++it){
+    Block * oBB = *it;
+    Block * newBB = gutils->getNewFromOriginal(oBB);
+    Block * reverseBB = gutils->mapReverseModeBlocks.lookupOrNull(oBB);
+
+    mapInvertArguments(oBB, reverseBB, gutils);
+
+    handleReturns(oBB, newBB, reverseBB, gutils);
+    
+    visitChildren(oBB, reverseBB, gutils);
+    
+    handlePredecessors(oBB, reverseBB, gutils);
+  }
+
+  auto nf = gutils->newFunc;
+
+  llvm::errs() << "nf\n";
+  nf.dump();
+  llvm::errs() << "nf end\n";
+  delete gutils;
```

Comment:
It is a very dangerous pattern to have `new` and `delete` in separate places. It is significantly better to return an `std::unique_ptr` from the op that creates this object, rather than a raw pointer that requires deletion.